kvlds-undump/main.c

```diff
@@ -0,0 +1,300 @@
+	C->N++;
+
+	/* Success! */
+	return (0);
+
+nomore:
+	/* No more key-value pairs; but we were successful. */
+	*key = *value = NULL;
+	return (0);
+
+err1:
+	kvldskey_free(*key);
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+static void
+usage(void)
+{
+
+	fprintf(stderr, "usage: kivaloo-kvlds-undump -t <kvlds socket>"
+	    " [--fs <dir>]\n");
+	fprintf(stderr, "       kivaloo-kvlds-undump --version\n");
+	exit(1);
+}
+
+/* Macro to simplify error-handling in command-line parse loop. */
+#define OPT_EPARSE(opt, arg) do {					\
+	warnp("Error parsing argument: %s %s", opt, arg);		\
+	exit(1);							\
+} while (0)
+
+int
+main(int argc, char * argv[])
+{
+	/* State variables. */
+	void * K;
+	struct wire_requestqueue * Q;
+	struct undumpstate C;
+	struct timeval st, en;
+
+	/* Command-line parameters. */
+	char * opt_fs = NULL;
+	char * opt_t = NULL;
+	int opt_v = 0;
+
+	/* Working variables. */
+	const char * ch;
+
+	WARNP_INIT;
+
+	/* Parse the command line. */
+	while ((ch = GETOPT(argc, argv)) != NULL) {
+		GETOPT_SWITCH(ch) {
+		GETOPT_OPTARG("--fs"):
+			if (opt_fs != NULL)
+				usage();
+			if ((opt_fs = strdup(optarg)) == NULL)
+				OPT_EPARSE(ch, optarg);
+			break;
+		GETOPT_OPTARG("-t"):
+			if (opt_t != NULL)
+				usage();
+			if ((opt_t = strdup(optarg)) == NULL)
+				OPT_EPARSE(ch, optarg);
+			break;
+		GETOPT_OPTARG("-v"):
+			opt_v++;
+			break;
+		GETOPT_OPT("--version"):
+			fprintf(stderr, "kivaloo-kvlds-undump @VERSION@\n");
+			exit(0);
+		GETOPT_MISSING_ARG:
+			warn0("Missing argument to %s\n", ch);
+			usage();
+		GETOPT_DEFAULT:
+			warn0("illegal option -- %s\n", ch);
+			usage();
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	/* We should have processed all the arguments. */
+	if (argc != 0)
+		usage();
+	(void)argv; /* argv is not used beyond this point. */
+
+	/* Sanity-check options. */
+	if (opt_t == NULL)
+		usage();
+
+	/* Open a connection to KVLDS. */
+	if ((K = kivaloo_open(opt_t, &Q)) == NULL) {
+		warnp("Could not connect to KVLDS daemon");
+		exit(1);
+	}
+
+	/*
+	 * If we're reading from the filesystem, move to that directory and
+	 * start reading it.
+	 */
+	if (opt_fs) {
+		if (chdir(opt_fs)) {
+			warnp("chdir(%s)", opt_fs);
+			exit(1);
+		}
+		if ((C.d = opendir(".")) == NULL) {
+			warnp("opendir(.)");
+			exit(1);
+		}
+	}
+
+	/* Prepare for undumping key-value pairs. */
+	C.N = 0;
+	if (!opt_fs)
+		C.d = NULL;
+
+	/* Get timestamp. */
+	if (monoclock_get(&st)) {
+		warnp("monoclock_get");
+		exit(1);
+	}
+
+	/* Store many key-value pairs. */
+	if (kvlds_multiset(Q, callback_pair, &C)) {
+		warnp("Error occurred while writing key-value pairs");
+		exit(1);
+	}
+
+	/* Get timestamp. */
+	if (monoclock_get(&en)) {
+		warnp("monoclock_get");
+		exit(1);
+	}
+
+	/* Print statistics if appropriate. */
+	if (opt_v) {
+		fprintf(stderr,
+		    "Stored %" PRIu64 " key-value pairs in %f seconds.",
+		    C.N, timeval_diff(en, st));
+	}
+
+	/* Close the directory (if applicable). */
+	if (opt_fs) {
```

Comment:
Style: Unnecessary braces.

Reply:
Solved: added stuff to make the braces not unnecessary.